rust-src/id/src/id_verifier.rs

```diff
@@ -0,0 +1,173 @@
+use crate::{types::*, utils};
+use bulletproofs::range_proof::{verify_efficient, Generators, RangeProof};
+use curve_arithmetic::Curve;
+use ff::Field;
+use pedersen_scheme::{
+    Commitment, CommitmentKey as PedersenKey, Randomness as PedersenRandomness, Value,
+};
+use random_oracle::RandomOracle;
+use sha2::{Digest, Sha256};
+
+/// Function for opening an attribute inside a commitment. The arguments are
+/// - keys - the commitments keys used to commit to the attribute
+/// - attribute - the attribute claimed to be inside the commitment
+/// - r - the randomness used to commit
+/// - c - the commitment
+///
+/// The function outputs a bool, indicicating whether the commitment contains
+/// the given attribute.
+pub fn verify_attribute<C: Curve, AttributeType: Attribute<C::Scalar>>(
+    keys: &PedersenKey<C>,
+    attribute: &AttributeType,
+    r: &PedersenRandomness<C>,
+    c: &Commitment<C>,
+) -> bool {
+    let s = Value::new(attribute.to_field_element());
+    keys.open(&s, &r, &c)
+}
+
+/// Function for verifying a range proof about an attribute inside a commitment.
+/// The arguments are
+/// - keys - the commitments keys used to commit to the attribute
+/// - gens - the bulletproof generators needed for range proofs
+/// - lower - the lower bound of the range
+/// - upper - the upper bound of the range
+/// - c - the commitment to the attribute
+/// - proof - the range proof about the attribute inside the commitment
+///
+/// The function outputs a bool, indicating whether the proof is correct or not,
+/// i.e., wether is attribute inside the commitment lies in [lower,upper).
+pub fn verify_attribute_range<C: Curve, AttributeType: Attribute<C::Scalar>>(
+    keys: &PedersenKey<C>,
+    gens: &Generators<C>,
+    lower: &AttributeType,
+    upper: &AttributeType,
+    c: &Commitment<C>,
+    proof: &RangeProof<C>,
+) -> bool {
+    let mut transcript = RandomOracle::domain("attribute_range_proof");
+    let a = lower.to_field_element();
+    let b = upper.to_field_element();
+    let zero_randomness = PedersenRandomness::<C>::new(C::Scalar::zero());
+    let com_a = keys.hide_worker(&a, &zero_randomness);
+    let com_b = keys.hide_worker(&b, &zero_randomness);
+    let two = C::scalar_from_u64(2);
+    let two_n = two.pow(&[64]);
+    let com_2n = keys.hide_worker(&two_n, &zero_randomness);
+    let com_delta_minus_b_plus_2n = Commitment(c.0.minus_point(&com_b.0).plus_point(&com_2n.0));
+    let com_delta_minus_a = Commitment(c.0.minus_point(&com_a.0));
+
+    verify_efficient(
+        &mut transcript,
+        64,
+        &[com_delta_minus_b_plus_2n, com_delta_minus_a],
+        proof,
+        gens,
+        keys,
+    )
+    .is_ok()
```

Comment:
Maybe return the Result instead. I don't think there is harm in that, and it might be useful in some cases to know why the proof failed (even though it is not very informative at the moment).